File: LeetCode/src/hashtable/validSudoku/ValidSudoku.java
package hashtable.validSudoku;

import java.util.Arrays;

import org.junit.Test;

public class ValidSudoku {
	public boolean isValidSudoku(char[][] board) {
		boolean[] used = new boolean[9];

		// row
		for (int row = 0; row < 9; ++row) {
			Arrays.fill(used, false);
			for (int i = 0; i < 9; ++i) {
				if (used(used, board[row][i]))
					return false;
			}
		}

		// column
		for (int col = 0; col < 9; ++col) {
			Arrays.fill(used, false);
			for (int i = 0; i < 9; ++i) {
				if (used(used, board[i][col]))
					return false;
			}
		}

		// sub matrix
		for (int i = 0; i < 9; i += 3) {
			for (int j = 0; j < 9; j += 3) {
				Arrays.fill(used, false);
				for (int k = 0; k < 9; k++) {
					if (used(used, board[i + k / 3][j + k % 3]))
						return false;
				}
			}
		}
		return true;
	}

	private boolean used(boolean[] used, char digit) {
		if (digit == '.')
			return false;
		int num = digit - '0' - 1;
		if (used[num])
			return true;
		used[num] = true;
		return false;
	}

	@Test
	public void test() {
		char[][] board = { 
				{ '.', '.', '4', '.', '.', '.', '6', '3', '.' },
				{ '.', '.', '.', '.', '.', '.', '.', '.', '.' }, 
				{ '5', '.', '.', '.', '.', '.', '.', '9', '.' },
				{ '.', '.', '.', '5', '6', '.', '.', '.', '.' }, 
				{ '4', '.', '3', '.', '.', '.', '.', '.', '1' },
				{ '.', '.', '.', '7', '.', '.', '.', '.', '.' }, 
				{ '.', '.', '.', '5', '.', '.', '.', '.', '.' },
				{ '.', '.', '.', '.', '.', '.', '.', '.', '.' }, 
				{ '.', '.', '.', '.', '.', '.', '.', '.', '.' } };
		System.out.println(isValidSudoku(board));
	}
}

File: LeetCode/src/hashtable/validSudoku/Valid_Sudoku.cpp
#include <vector>
#include <iostream>
using std::vector;
using std::fill;

bool is_used(bool used[], char& c) {
	if (c == '.')
		return false;
	int num = c - '0' - 1;
	if (used[num])
		return true;
	used[num] = true;
	return false;
}

bool isValidSudoku(vector<vector<char>>& board) {
	bool used[9];

	for (int i = 0; i < 9; ++i) {
		//row
		fill(used, used + 9, false);
		for (int j = 0; j < 9; ++j) {
			if (is_used(used, board[i][j]))
				return false;
		}

		//col
		fill(used, used + 9, false);
		for (int k = 0; k < 0; ++k) {
			if (is_used(used, board[k][i]))
				return false;
		}
	}

	for (int i = 0; i < 9; i += 3) {
		for (int j = 0; j < 9; j += 3) {
			fill(used, used + 9, false);
			for (int k = 0; k < 9; ++k) {
				if (is_used(used, board[i + k / 3][j + k % 3]))
					return false;
			}
		}
	}
	return true;
}